platform/engineering/design-docs/2020-09-27-broken-links.md

```diff
@@ -0,0 +1,122 @@
+
+The CMS team built the [Node Link Report](https://www.drupal.org/project/node_link_report) module (open sourced module) to detect and warn of broken links upon node save. Editors now have immediate feedback about potential broken links before publishing. This greatly reduced the number of broken link related content release stoppages.  We have done as much as we can on that front at the CMS layer.
+
+
+### Guiding Principles
+1. Veterans should not get 404 errors. Content on VA.gov should not have broken links.
+2. As the CMS scales from ~20 editors and daily content releases, up to ~500 editors and continuous content releases, we can not allow a broken link to prevent content releases.  Knowing about a storm closure or active shooter [saves more souls](https://www.informationweek.com/government/dont-get-side-tracked-by-flashy-metrics-success-is-in-the-people/d/d-id/1331791) than not getting a 404 when following a link.
+3. Developers currently waste far too much time chasing down broken links.
+
+
+### High Level Design
+Currently the Front End link checker creates a CSV file and exposes that to Jenkins.  The CSV contains the location of the broken link, the url it is pointing to.  I propose publishing that file to the website as open data (perhaps located at /data/vagov-broken-links.csv or as json).  Currently the Front End build process notifies the CMS when a content release has succeeded by pinging the CMS  The CMS could act on that ping as a trigger and read the broken link data.  It could process it immediately and take the following actions:
+1. Look up the location of the broken link, and determine if it is a CMS node.
+2. If it is a CMS node, notify the editor that created the published revision immediately in the CMS and send an email to them.
+    1. If the same page has a broken link for 3 separate content releases  _(~15 min span of time)_ notify other content owners responsible for the same section.
+    2. If the same page has a broken link for 5 separate content releases _(~25 min span of time)_, or there are no other editors for that section, alert Slack #CMS-team so the team can escalate the fix.
+3. If it is not a CMS based page, notify the appropriate Slack channel based on the location of the broken link.
+4. Sections that are under construction (hidden behind basic auth) could be disregarded so broken links in that section would be ignored by the CMS to avoid creating extra noise.
+
+## Specifics
+
+### Detailed Design
+There are three phases of changes that would need to be implemented across the two systems in the following order.
+
+1. Front end build makes broken link CSV public.  The CSV already gets built, it just needs to be saved as a public file.  (https://www.va.gov/data/vagov-broken-links.csv)
```

Comment:
I think if this process is not going to block content deployments because of broken links, then the build process doesn't have to be part of this validation. The current production html can be synced and run against an independent link checker like [broken-link-checker](https://www.npmjs.com/package/broken-link-checker) or [link-checker](https://packagist.org/packages/glicer/link-checker) and the validation can be triggered by successful deployment or on a schedule. 

What do you think?

Reply:
That's an interesting point.  Would this suggestion gain us speed improve the timing? It would have to be fast, and the broken link file that is exposed would have to match the current release.  Otherwise it could create wild goose chases for a broken link report that is out of sync with the actual files.
I'd rather not increase the complexity and change the current method if it doesn't gain us anything.    Right now the csv is produced, it just needs to be made public.  If platform wants to swap one link checker method for another that would pose no problem as long as that method could be in sync with the current release and have the csv ready when the front end release pings drupal to day the release is complete.

Reply:
>It would have to be fast, and the broken link file that is exposed would have to match the current release.

Can we add more information about matching the current release to the design doc?